### PR TITLE
[ANDROID] Support for passing in an Application directly to the AndroidSentryClientFactory

### PR DESCRIPTION
**Pull request by justacodefan**
Hopefully resolves the use case for #713

### CODE REVIEW

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 1]
```diff
@@ -30,22 +30,38 @@
...
      */
     public AndroidSentryClientFactory(Context ctx) {
-        Log.d(TAG, "Construction of Android Sentry.");
+        Log.d(TAG, "Construction of Android Sentry from Android Context.");
+
+        this.ctx = ctx;
+    }
 
-        this.ctx = ctx.getApplicationContext();
+    public Context getAndroidContext() {
```
Comment (ninniuz) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: What is the purpose of this method?

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 2]
```diff
@@ -30,22 +30,38 @@
...
     /**
      * Construct an AndroidSentryClientFactory using the specified Android Context.
      *
      * @param ctx Android Context.
      */
     public AndroidSentryClientFactory(Context ctx) {
-        Log.d(TAG, "Construction of Android Sentry.");
+        Log.d(TAG, "Construction of Android Sentry from Android Context.");
+
+        this.ctx = ctx;
```
Comment (ninniuz) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: This change is a bit risky, as it is going to produce a memory leak if an application is creating a `AndroidSentryClientFactory` with an `Activity` as `ctx`.

Sentry would hold a reference to such `Activity` and any resource tied to it won't be eligible for garbage collection (e.g. the whole view hierarchy, which can be massive in terms of memory occupation).

I am not saying that the previous method of getting hold of a Context was wrong, I just suggested a way for a developer to feed consciously a Context to be used by Sentry.

File: sentry/src/main/java/io/sentry/DefaultSentryClientFactory.java [hunk 3]
```diff
@@ -535,7 +535,7 @@ protected ContextManager getContextManager(Dsn dsn) {
             }
         }
 
-        return inAppPackages;
+        return Collections.unmodifiableList(inAppPackages);
```
Comment (ninniuz) [inline sentry/src/main/java/io/sentry/DefaultSentryClientFactory.java]: I think this kind of breaks the current API contract in a subtle way: while this is ABI compatible, subclasses will fail at runtime trying to add up to the base class collection of in-app packages.

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java (hunk 2 above)
Comment (justacodefan) [reply sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: The factories are not long lived and in fact they should not live any longer than it is required to create a new `SentryClient`. I can also wrap `ctx` in a `WeakReference` however.

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java (hunk 1 above)
Comment (justacodefan) [reply sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: That needs to be inlined; I was initially going with a different approach and just didn't remove it yet.

**Comment by justacodefan** [comment]
@ninniuz could you take another look please?

**Comment by ninniuz** [comment]
@justacodefan that works 👍 

Do you think it is feasible to make `io.sentry.DefaultSentryClientFactory#getInAppFrames` return a mutable list instead of `Collections.emptyList()` if there is no in-app frames property?

It is kind of odd to receive either a mutable or immutable list as a result of the same method call.

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 4]
```diff
@@ -14,9 +16,9 @@
 import io.sentry.context.SingletonContextManager;
 import io.sentry.dsn.Dsn;
 import io.sentry.util.Util;
-
 import java.io.File;
 import java.util.ArrayList;
+import java.lang.ref.WeakReference;
```
Comment (ninniuz) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: I think this is now unused and could be removed.

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 5]
```diff
@@ -30,32 +32,48 @@
...
+
     @Override
     public SentryClient createSentryClient(Dsn dsn) {
         if (!checkPermission(Manifest.permission.INTERNET)) {
             Log.e(TAG, Manifest.permission.INTERNET + " is required to connect to the Sentry server,"
                 + " please add it to your AndroidManifest.xml");
         }
 
-        Log.d(TAG, "Sentry init with ctx='" + ctx.toString() + "'");
+        Log.d(TAG, "Sentry init with ctx='" + this.ctx.toString() + "'");
```
Comment (ninniuz) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: // nitpick-mode-on
Can you change `this.ctx` to `ctx`?
// nitpick-mode-off

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 6]
```diff
@@ -73,7 +91,7 @@ public SentryClient createSentryClient(Dsn dsn) {
         }
 
         SentryClient sentryClient = super.createSentryClient(dsn);
-        sentryClient.addBuilderHelper(new AndroidEventBuilderHelper(ctx));
+        sentryClient.addBuilderHelper(new AndroidEventBuilderHelper(this.ctx));
```
Comment (ninniuz) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: // nitpick-mode-on
Can you change `this.ctx` to `ctx`?
// nitpick-mode-off

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 7]
```diff
@@ -85,7 +103,7 @@ public SentryClient createSentryClient(Dsn dsn) {
         if (inAppFrames.isEmpty()) {
             PackageInfo info = null;
             try {
-                info = ctx.getPackageManager().getPackageInfo(ctx.getPackageName(), 0);
+                info = this.ctx.getPackageManager().getPackageInfo(this.ctx.getPackageName(), 0);
```
Comment (ninniuz) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: // nitpick-mode-on
Can you change `this.ctx` to `ctx`?
// nitpick-mode-off

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 8]
```diff
@@ -107,7 +125,7 @@ protected Buffer getBuffer(Dsn dsn) {
         if (bufferDirOpt != null) {
             bufferDir = new File(bufferDirOpt);
         } else {
-            bufferDir = new File(ctx.getCacheDir().getAbsolutePath(), DEFAULT_BUFFER_DIR);
+            bufferDir = new File(this.ctx.getCacheDir().getAbsolutePath(), DEFAULT_BUFFER_DIR);
```
Comment (ninniuz) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: // nitpick-mode-on
Can you change `this.ctx` to `ctx`?
// nitpick-mode-off

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 9]
```diff
@@ -123,11 +141,11 @@ protected ContextManager getContextManager(Dsn dsn) {
      * Check whether the application has been granted a certain permission.
      *
      * @param permission Permission as a string
+     *
      * @return true if permissions is granted
      */
     private boolean checkPermission(String permission) {
-        int res = ctx.checkCallingOrSelfPermission(permission);
+        int res = this.ctx.checkCallingOrSelfPermission(permission);
```
Comment (ninniuz) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: // nitpick-mode-on
Can you change `this.ctx` to `ctx`?
// nitpick-mode-off

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java (hunk 6 above)
Comment (bruno-garcia) [reply sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: sure

**Pull request merged**